syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.rubble.logAndApply";
option java_outer_classname = "logAndApplyProto";
option objc_class_prefix = "lAA";

package logAndApply;

// The logAndApply service.
service logAndApply {
    rpc logAndApply (EditLists) returns (Response){}
}

message Response {
  bool ok = 1;
}

message EditLists {
  message EditList {
    message VersionEdit {
      NewFiles added = 1;
      DeleteFiles dels = 2;
    }
    repeated VersionEdit version_edit = 1;
  }  
  repeated EditList edit_list = 1;
}

message NewFile{
  int32 level = 1;

  message FileMetaData {
    message FileDescriptor {
      uint64  file_number = 1;
      uint64  file_size = 2;
      // probably need to send seqno
      optional uint64 smallest_seqno = 3;
      optional uint64 largest_seqno = 4;
    }  
    FileDescriptor fd = 1;
    string  smallest_key = 2;
    string  largest_key = 3;

    // not sure if needed to send those fields
    optional uint64 oldest_ancestor_time = 4;
    optional uint64 file_creation_time = 5;
    optional string file_checksum = 6;
    optional string file_checksum_func_name = 7;
  }
  FileMetaData meta = 2;
}

message NewFiles{
  repeated NewFile new = 1;
}

message DeleteFile {
  int32   level = 1;
  uint64  file_number = 2;
}

message DeleteFiles {
  repeated DeleteFile del = 1;
}



